DigitalLearningSolutions.Data/Services/NotificationService.cs

```diff
@@ -1,61 +1,74 @@
+            IConfigService configService,
+            IEmailService emailService,
+            IFeatureManager featureManager
+        )
+        {
+            this.notificationDataService = notificationDataService;
+            this.configService = configService;
+            this.emailService = emailService;
+            this.featureManager = featureManager;
+        }
+
+        public void SendUnlockRequest(int progressId)
+        {
+            var unlockData = notificationDataService.GetUnlockData(progressId);
+            if (unlockData == null)
+            {
+                throw new NotificationDataException(
+                    $"No record found when trying to fetch Unlock Data. Progress ID: {progressId}"
+                );
+            }
+
+            unlockData.ContactForename = unlockData.ContactForename == "" ? "Colleague" : unlockData.ContactForename;
+            var refactoredTrackingSystemEnabled = featureManager.IsEnabledAsync("RefactoredTrackingSystem").Result;
+            var baseUrlConfigOption = refactoredTrackingSystemEnabled ? ConfigService.AppBaseUrl : ConfigService.TrackingSystemBaseUrl;
+            var trackingSystemBaseUrl = configService.GetConfigValue(baseUrlConfigOption) ??
+                                        throw new ConfigValueMissingException(
+                                            configService.GetConfigValueMissingExceptionMessage(baseUrlConfigOption)
+                                        );
+            var unlockUrl = new UriBuilder(trackingSystemBaseUrl);
+
+            unlockUrl.Path += "coursedelegates";
+            unlockUrl.Query = $"CustomisationID={unlockData.CustomisationId}";
+            string emailSubjectLine = "Digital Learning Solutions Progress Unlock Request";
+            var builder = new BodyBuilder
+            {
+                TextBody = $@"Dear {unlockData?.ContactForename}
+Digital Learning Solutions Delegate, {unlockData?.DelegateName}, has requested that you unlock their progress for the course {unlockData?.CourseName}.
+They have reached the maximum number of assessment attempt allowed without passing.
+To review and unlock their progress, visit the this url: ${unlockUrl.Uri}.",
+                HtmlBody = $@"<body style= 'font-family: Calibri; font-size: small;'>
+                                    <p>Dear {unlockData?.ContactForename}</p>
+                                    <p>Digital Learning Solutions Delegate, {unlockData?.DelegateName}, has requested that you unlock their progress for the course {unlockData?.CourseName}</p>
+                                    <p>They have reached the maximum number of assessment attempt allowed without passing.</p><p>To review and unlock their progress, <a href='{unlockUrl.Uri}'>click here</a>.</p>
+                                </body>",
+            };
+
+            emailService.SendEmail(
+                new Email(emailSubjectLine, builder, unlockData.ContactEmail, unlockData.DelegateEmail)

```

Comment:
There's several instances of null access in this chunk of code, the first of which will show as unnecessary, and these cause the access unlockData.ContactEmail to be marked as deference of a null reference.

All of these can be removed since we check unlockData == null at the top of the method.